File: Anul 3/Sem1/Implementarea Concurentei in Limbaje de Programare (optional)/Laborator/go/callable_future_task/callable_future_task.go
package main

import (
	"fmt"
	"time"
)

// Helper function to simulate a thread-like behavior
func runInThread(fn func()) {
	go fn()
}

// Equivalent to Callable<String> from Java
// Instead of an interface, we can use a function type func() (string, error)
// or just a channel to receive the result.
func MyTask() (string, error) {
	time.Sleep(5 * time.Second)
	return "Task completed", nil
}

func main() {
	// Java: Callable<String> task = new MyTask();
	// In Go, we just refer to the function or a closure
	task := MyTask

	// Java: FutureTask<String> futureTask = new FutureTask<>(task);
	// Java: Thread thread = new Thread(futureTask);
	// Java: thread.start();

	// In Go, we launch a goroutine that pushes to a channel
	// to simulate the "Future" part.
	resultChan := make(chan string)
	errChan := make(chan error)

	go func() {
		res, err := task()
		if err != nil {
			errChan <- err
		} else {
			resultChan <- res
		}
	}()

	fmt.Println("Main...")

	// Java: String result = futureTask.get();
	// This blocks until the result is available
	select {
	case res := <-resultChan:
		fmt.Println("Result: " + res)
	case err := <-errChan:
		fmt.Printf("Error: %v\n", err)
	}
}

File: Anul 3/Sem1/Implementarea Concurentei in Limbaje de Programare (optional)/Laborator/java/CallableFutureTask.java
import java.util.concurrent.Callable;
import java.util.concurrent.FutureTask;

public class CallableFutureTask {
    public static void main(String[] args) {
        Callable<String> task = new MyTask();

        // sau cu lambda
        // Callable<String> task = () -> {
        //     Thread.sleep(50000);
        //     return "Task completed";
        // };

        FutureTask<String> futureTask = new FutureTask<>(task);

        Thread thread = new Thread(futureTask);
        thread.start();

        try {
            System.out.println("Main...");
            String result = futureTask.get();
            System.out.println("Result: " + result);
        }
        catch (Exception ex) {
            ex.printStackTrace();
        }
    }
}

class MyTask implements Callable<String> {
    @Override
    public String call() {
        try {
            Thread.sleep(5000);
            return "Task completed";
        }
        catch (InterruptedException ex) {
            ex.printStackTrace();
            return "Error";
        }
    }
}